### ISSUE TITLE
Multiprocessing Data Generation Performance Bottleneck

### ISSUE THREAD

**Issue by zachtheyek** [zachtheyek/Aetherscan#15]
## Summary

Data generation in the training pipeline takes approximately **160 minutes per round** on a 96-core system, creating a severe bottleneck that dominates total training time. Additionally, rounds 2+ exhibit a **3-5x slowdown** compared to round 1, despite identical workloads.

## Environment

- **CPU**: 96 cores (dual-socket Xeon or similar)
- **RAM**: 500 GB
- **GPUs**: 6× 16GB (used for VAE training, not data generation)
- **Python**: 3.10+
- **TensorFlow**: 2.x (GPU build)
- **Key dependencies**: setigen, numpy, scipy

### Primary: Excessive Data Generation Time

Each training round calls \`DataGenerator.generate_batch()\` which uses \`multiprocessing.Pool\` to parallelize cadence creation across 96 workers. Despite high core count, data generation takes ~160 minutes per round.

Observed timing breakdown per chunk (10,000 cadences):
\`\`\`
Chunk processing: ~2 minutes
Total chunks per round: ~80
Total time: ~160 minutes
\`\`\`

### Secondary: Round 2+ Performance Degradation

| Round | Data Generation Time | Relative Performance |
|-------|---------------------|---------------------|
| 1     | ~45 minutes         | Baseline            |
| 2     | ~160 minutes        | 3.5x slower         |
| 3+    | ~160-180 minutes    | 3.5-4x slower       |

The slowdown persists even after calling \`tf.keras.backend.clear_session()\` between rounds.

## Root Cause Analysis

We identified four independent bottlenecks contributing to the performance issues:

### 1. Excessive Task Dispatch Overhead

**Current behavior**: \`pool.imap()\` is called with \`chunksize=1\`, resulting in 120,000 individual task dispatches per round.

\`\`\`python
# Current implementation
for i, result in enumerate(pool.imap(_single_cadence_wrapper, args_list, chunksize=1)):
    cadence[i] = result
\`\`\`

**Impact**: Each task dispatch involves:
- Lock acquisition on the task queue
- Pickle serialization of arguments (~1KB per task)
- Worker thread wake-up signaling
- Lock acquisition for result retrieval

With 120,000 dispatches, scheduling overhead dominates actual computation time.

**Evidence**: Profiling shows workers spending >60% of time in \`queue.get()\` and \`queue.put()\` rather than cadence generation.

### 2. IPC Data Transfer Bottleneck (Dominant Factor)

**Current behavior**: Each worker returns a 192KB numpy array (shape: \`16×256×1×3\`, dtype: \`float32\`) through the multiprocessing result queue.

\`\`\`python
# Worker returns full array through IPC
def _single_cadence_wrapper(args):
    # ... generate cadence ...
    return cadence  # 192KB array, pickled and sent through pipe
\`\`\`

**Impact**: 
- Per round: 120,000 results × 192KB = **~23GB of pickle serialization**
- Pickle \`dumps()\` and \`loads()\` are pure Python, requiring GIL for entire duration
- Data flows through OS pipes with additional kernel overhead

**Evidence**: \`py-spy\` profiling shows main process spending majority of time in \`pickle.loads()\` during \`pool.imap()\` iteration.

### 3. Sequential Synchronization Barriers

**Current behavior**: \`generate_batch()\` makes 8 sequential calls to \`batch_create_cadence()\`, each with its own synchronization barrier:

\`\`\`python
def generate_batch(self, num_samples, snr_base, snr_range):
    # 8 sequential pool operations, each blocking until complete
    main_quarter_1 = batch_create_cadence(pool, ...)  # barrier 1
    main_quarter_2 = batch_create_cadence(pool, ...)  # barrier 2
    main_quarter_3 = batch_create_cadence(pool, ...)  # barrier 3
    main_quarter_4 = batch_create_cadence(pool, ...)  # barrier 4
    false_half_1 = batch_create_cadence(pool, ...)    # barrier 5
    false_half_2 = batch_create_cadence(pool, ...)    # barrier 6
    true_half_1 = batch_create_cadence(pool, ...)     # barrier 7
    true_half_2 = batch_create_cadence(pool, ...)     # barrier 8
\`\`\`

**Impact**: 
- Workers sit idle while waiting for slowest task in each batch
- 8 synchronization points compound tail latency effects
- Cannot overlap work across different cadence types

### 4. TensorFlow Thread Pool GIL Contention (Causes Round 2+ Degradation)

**Current behavior**: TensorFlow initializes inter-op and intra-op thread pools on first operation. These pools persist for the process lifetime and cannot be paused or destroyed.

\`\`\`python
# In TrainingPipeline.__init__()
with self.strategy.scope():
    self.vae = create_beta_vae_model()  # TF runtime initializes here
    # Creates 96 inter-op threads + 96 intra-op threads
    # These 192 threads persist FOREVER
\`\`\`

**Impact on Round 1**:
- TF thread pools just created, minimal polling activity
- GIL contention during pickle operations is low
- Data generation completes in ~45 minutes

**Impact on Rounds 2+**:
- TF thread pools have been "warmed up" by training
- 192 threads continuously polling for work, periodically acquiring GIL
- Main process pickle operations constantly interrupted

\`\`\`
Round 1 GIL timeline:
Main:    [====pickle====][====pickle====][====pickle====]
TF:      [sleep..............................]

Round 2+ GIL timeline:  
Main:    [==pk][wait][==le][wait][==pk][wait][==le]...
TF(1):        [poll]      [poll]      [poll]
TF(2):              [poll]      [poll]      [poll]
...
TF(192):   [poll]   [poll]   [poll]   [poll]   [poll]
\`\`\`

**Evidence**: 
- Round 1 vs Round 2 timing difference is reproducible
- \`tf.config.threading.set_inter_op_parallelism_threads()\` fails with "cannot be modified after initialization"
- Slowdown does not occur if TensorFlow is never imported

### Solution 1: Shared Memory Output Arrays

Eliminate IPC pickle overhead by having workers write directly to pre-allocated shared memory.

\`\`\`python
from multiprocessing import shared_memory

def generate_batch(self, num_samples, snr_base, snr_range):
    # Pre-allocate output in shared memory
    output_shape = (total_cadences, 16, 256, 1, 3)
    output_shm = shared_memory.SharedMemory(create=True, size=np.prod(output_shape) * 4)
    output_array = np.ndarray(output_shape, dtype=np.float32, buffer=output_shm.buf)
    
    # Workers write directly to shared memory, return None
    # Zero pickle overhead for results
\`\`\`

**Expected impact**: 10-20x speedup by eliminating ~23GB pickle serialization per round.

### Solution 2: Coarse Task Granularity

Batch multiple cadences per task to amortize dispatch overhead.

\`\`\`python
def _batch_cadence_worker(task):
    """Process 500 cadences per task instead of 1."""
    batch_indices, output_offset, params = task
    for i, idx in enumerate(batch_indices):
        cadence = generate_single_cadence(idx, params)
        _GLOBAL_OUTPUT_ARRAY[output_offset + i] = cadence
    return None

# 120,000 tasks → 240 tasks (500x reduction)
batch_size = 500
tasks = [(indices[i:i+batch_size], i, params) for i in range(0, len(indices), batch_size)]
pool.map(_batch_cadence_worker, tasks)
\`\`\`

**Expected impact**: 2-5x speedup from reduced scheduling overhead.

### Solution 3: Unified Batch Submission

Submit all cadence types in a single \`pool.map()\` call instead of 8 sequential calls.

\`\`\`python
def generate_batch(self, num_samples, snr_base, snr_range):
    all_tasks = []
    all_tasks.extend(build_quarter_tasks(...))      # Was 4 separate calls
    all_tasks.extend(build_false_half_tasks(...))   # Was 2 separate calls  
    all_tasks.extend(build_true_half_tasks(...))    # Was 2 separate calls
    
    # Single dispatch, single synchronization point
    pool.map(_batch_cadence_worker, all_tasks)
\`\`\`

**Expected impact**: 1.5-2x speedup from eliminating 7 unnecessary sync barriers.

### Solution 4: Process Isolation for TensorFlow-Free Data Generation

Eliminate TF thread GIL contention by running data generation in a subprocess that never imports TensorFlow.

\`\`\`python
# Main process (has TensorFlow)
def generate_batch(self, num_samples, snr_base, snr_range):
    # Create shared memory for output
    output_shm = shared_memory.SharedMemory(create=True, size=output_nbytes)
    
    # Spawn subprocess that never imports TensorFlow
    config = {'output_shm_name': output_shm.name, 'num_samples': num_samples, ...}
    subprocess.run([sys.executable, '-m', 'aetherscan.isolated_data_gen', json.dumps(config)])
    
    # Read results from shared memory
    return np.ndarray(output_shape, buffer=output_shm.buf).copy()
\`\`\`

\`\`\`python
# isolated_data_gen.py - Subprocess entry point
# This module NEVER imports TensorFlow

assert 'tensorflow' not in sys.modules, "TF must not be imported!"

def run_data_generation(config):
    # Pure numpy/setigen operations
    # Zero TF threads = Zero GIL contention
    # Identical performance on all rounds
\`\`\`

**Expected impact**: Eliminates 3-5x round 2+ degradation; all rounds perform like round 1.

## Combined Performance Projection

| Optimization | Individual Speedup | Cumulative Time |
|--------------|-------------------|-----------------|
| Baseline (current) | - | ~160 min |
| + Shared memory outputs | 10-20x | ~10-15 min |
| + Coarse task granularity | 2-5x | ~3-5 min |
| + Unified submission | 1.5-2x | ~2-3 min |
| + Process isolation | Fixes round 2+ regression | ~2-3 min (all rounds) |

**Target**: Reduce data generation from ~160 minutes to <5 minutes per round, with consistent performance across all rounds.

### Phase 1: Core Optimizations (Highest Impact)
- [ ] Implement shared memory output arrays in \`DataGenerator\`
- [ ] Add \`_batch_cadence_worker()\` for coarse task granularity
- [ ] Refactor \`generate_batch()\` for unified task submission
- [ ] Add integration tests to verify output correctness

### Phase 2: Process Isolation
- [ ] Create \`isolated_data_gen.py\` subprocess module
- [ ] Implement shared memory handoff between processes
- [ ] Verify TensorFlow is never imported in subprocess
- [ ] Add error handling for subprocess failures

### Phase 3: Validation & Monitoring
- [ ] Add timing instrumentation to \`generate_batch()\`
- [ ] Compare round 1 vs round 2+ performance
- [ ] Monitor shared memory usage with \`ipcs -m\`
- [ ] Profile with \`py-spy\` to confirm IPC overhead eliminated

## References

- [Python multiprocessing shared memory documentation](https://docs.python.org/3/library/multiprocessing.shared_memory.html)
- [TensorFlow threading configuration](https://www.tensorflow.org/api_docs/python/tf/config/threading)
- [GIL contention in multiprocessing](https://docs.python.org/3/glossary.html#term-global-interpreter-lock)

---

**Suggested Labels**: \`performance\`, \`multiprocessing\`, \`training-pipeline\`, \`high-priority\`

**Comment by zachtheyek** [comment]
initial issue was written by claude code. fyi, i've not verified their diagnosis or solutions are accurate, so take with a pinch of salt

**Comment by zachtheyek** [comment]
@claude please triage and assign labels to this issue

**Comment by zachtheyek** [comment]
@claude please run the following github actions jobs on this issue: 
- auto-assign-author.yml
- claude-issue-triage.yml 
- claude-contribution-check.yml

**Comment by zachtheyek** [comment]
Closed as resolved by PR #117 (training-data-pipeline overhaul, merged in master `6e89608`): all four diagnosed mechanisms were designed out — per-sample task dispatch → batched tasks (`--data-gen-task-size`, default 64 after production-scale tuning on both clusters), serialization overhead → direct memmap writes, per-batch pool churn → one unified pool.map barrier, and the GIL-bound generation → a spawned background producer that generates round k+1 while round k trains. `generate_batch()`/`batch_create_cadence()` no longer exist in `data_generation.py`. Production sweeps at n=499200 measured 55.4 samples/s (96c) / 29.8 samples/s (32c). The stale KNOWN_ISSUES §8 entry that still points here flips to Closed via open PR #147. Remaining tail-latency work (create_true_double straggler) is tracked separately in #118.
